opencensus/trace/tracer.py

```diff
@@ -74,8 +74,7 @@ def should_sample(self):
         :rtype: bool
         :returns: Whether to trace the request or not.
         """
-        return self.span_context.trace_options.enabled \
-            or self.sampler.should_sample(self.span_context.trace_id)
+        return self.sampler.should_sample(self.span_context)
```

Comment:
This is the important change: we used to short-circuit here, which meant we couldn't refuse to sample (from e.g. the `AlwaysOffSampler`) if the sampling bit was already set.

Note that we're reading `span_context.trace_options.enabled` in the sampler... and then using the sampling decision to set it. This works because we only use the sampler once at init time to set `tracer.tracer` (and the sampling bit). This is potentially confusing, but refactoring `Tracer` to fix this is outside the scope of this PR.